In [5]:
library(tidyverse)
library(foreign) # stata data 
library(car) # companion to regressions
library(lmtest) # white test

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.3     ✔ purrr   0.3.4
✔ tibble  3.1.0     ✔ dplyr   1.0.5
✔ tidyr   1.1.3     ✔ stringr 1.4.0
✔ readr   1.4.0     ✔ forcats 0.5.1

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter()  masks stats::filter()
✖ purrr::flatten() masks jsonlite::flatten()
✖ dplyr::lag()     masks stats::lag()

Loading required package: carData


Attaching package: ‘car’


The following object is masked from ‘package:dplyr’:

    recode


The following object is masked from ‘package:purrr’:

    some


Loading required package: zoo


Attaching package: ‘zoo’


The following objects are masked from ‘package:base’:

    as.Date, as.Date.numeric




## Let `noPC` be a dummy variable equal to 1 if the student does not own a computer, and zero otherwise. 

In [6]:
# import the dataset from stata (.dta) using the foreign package
gpa1 <- read.dta("/Users/ama/OneDrive - usp.br/S3/Natalia/Topico 5/GPA1.DTA")

Warning message in read.dta("/Users/ama/OneDrive - usp.br/S3/Natalia/Topico 5/GPA1.DTA"):
“cannot read factor labels from Stata 5 files”


In [7]:
# the original regression
gpa1_first_lm <- lm(colGPA ~ PC + hsGPA + ACT, data = gpa1)

# results from the regression
summary(gpa1_first_lm)


Call:
lm(formula = colGPA ~ PC + hsGPA + ACT, data = gpa1)

Residuals:
    Min      1Q  Median      3Q     Max 
-0.7901 -0.2622 -0.0107  0.2334  0.7570 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 1.263520   0.333125   3.793 0.000223 ***
PC          0.157309   0.057287   2.746 0.006844 ** 
hsGPA       0.447242   0.093647   4.776 4.54e-06 ***
ACT         0.008659   0.010534   0.822 0.412513    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3325 on 137 degrees of freedom
Multiple R-squared:  0.2194,	Adjusted R-squared:  0.2023 
F-statistic: 12.83 on 3 and 137 DF,  p-value: 1.932e-07


### What happens to the intercept if `noPC` is used instead of `PC`?

The new intercept can be calculated with the $\hat{\beta}_{0}$ plus the $\hat{\delta}_{0}$ of the original regression, so 1.26 + .157 = 1.417, which is what we see when we run the regression. 

In [8]:
# same regression, but with the inverse of the dummy PC
gpa1_sec_lm <- lm(colGPA ~ I(1 - PC) + hsGPA + ACT, data = gpa1)

summary(gpa1_sec_lm)


Call:
lm(formula = colGPA ~ I(1 - PC) + hsGPA + ACT, data = gpa1)

Residuals:
    Min      1Q  Median      3Q     Max 
-0.7901 -0.2622 -0.0107  0.2334  0.7570 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept)  1.420829   0.336605   4.221 4.40e-05 ***
I(1 - PC)   -0.157309   0.057287  -2.746  0.00684 ** 
hsGPA        0.447242   0.093647   4.776 4.54e-06 ***
ACT          0.008659   0.010534   0.822  0.41251    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3325 on 137 degrees of freedom
Multiple R-squared:  0.2194,	Adjusted R-squared:  0.2023 
F-statistic: 12.83 on 3 and 137 DF,  p-value: 1.932e-07


### What happens to the R^2 if `noPC` is used instead of `PC`?

Nothing happens to the R-squared because using `noPC` is just a different way of representing the same information about computer ownership. 

### Should `PC` and `noPC` both be included as independent variables in the model?

Both dummy variables should not be included in the regression becaue the information is redundant and the two variables will be perfectly correlated, which would invalidate the model.

## Use the data in GPA1.RAW for this exercise.

### What happens to the estimated effect of PC ownership with added `mothcoll` and `fathcoll` ? Is PC still statistically significant?

The eeffect of `PC` changes only slightly, and remains statistically significant. 

In [9]:
gpa1_third_lm <- lm(colGPA ~ PC + hsGPA + ACT + mothcoll + fathcoll, data = gpa1)

summary(gpa1_third_lm)


Call:
lm(formula = colGPA ~ PC + hsGPA + ACT + mothcoll + fathcoll, 
    data = gpa1)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.78149 -0.25726 -0.02121  0.24691  0.74432 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept)  1.255554   0.335392   3.744 0.000268 ***
PC           0.151854   0.058716   2.586 0.010762 *  
hsGPA        0.450220   0.094280   4.775 4.61e-06 ***
ACT          0.007724   0.010678   0.723 0.470688    
mothcoll    -0.003758   0.060270  -0.062 0.950376    
fathcoll     0.041800   0.061270   0.682 0.496265    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3344 on 135 degrees of freedom
Multiple R-squared:  0.2222,	Adjusted R-squared:  0.1934 
F-statistic: 7.713 on 5 and 135 DF,  p-value: 2.083e-06


### Joint significance of `mothcoll` and `fathcoll`.

The F test for the joint significance of `mothcoll` and `fathcoll` is about .24 with a p-value of .78, meaning that they are jointly quite insignificant. This explains why the coefficients of other variables didn't change much when we added these two variables to the regression. 

In [10]:
# impose the restrictions of the null by setting the variables to zero 

# gpa1_first_lm is the same as the restricted model
summary(gpa1_first_lm) 


Call:
lm(formula = colGPA ~ PC + hsGPA + ACT, data = gpa1)

Residuals:
    Min      1Q  Median      3Q     Max 
-0.7901 -0.2622 -0.0107  0.2334  0.7570 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 1.263520   0.333125   3.793 0.000223 ***
PC          0.157309   0.057287   2.746 0.006844 ** 
hsGPA       0.447242   0.093647   4.776 4.54e-06 ***
ACT         0.008659   0.010534   0.822 0.412513    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3325 on 137 degrees of freedom
Multiple R-squared:  0.2194,	Adjusted R-squared:  0.2023 
F-statistic: 12.83 on 3 and 137 DF,  p-value: 1.932e-07


In [11]:
 # calculate the F statistic by hand
((0.2222 - 0.2194) / (137 - 135)) / ((1 - 0.2222) / (141 - 4 - 1))

[1] 0.244793

In [12]:
# calculate the F statistic with the car package

# define the null hypothesis
null_hyp <- c("mothcoll", "fathcoll")  

linearHypothesis(gpa1_third_lm, null_hyp)

Res.Df RSS      Df Sum of Sq  F         Pr(>F)   
1 137    15.14868 NA         NA        NA        NA
2 135    15.09400  2 0.05468536 0.2445517 0.7834017

### Add `hsGPA^2` and decide whether this generalization is needed.

`hsGPA^2` has a coefficient of about .337 and a p-value of .12, 

In [13]:
gpa1_squared_lm <- lm(colGPA ~ PC + hsGPA + I(hsGPA^2) + ACT + mothcoll + fathcoll, data = gpa1)

summary(gpa1_squared_lm)


Call:
lm(formula = colGPA ~ PC + hsGPA + I(hsGPA^2) + ACT + mothcoll + 
    fathcoll, data = gpa1)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.78998 -0.24327 -0.00648  0.26179  0.72231 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)  
(Intercept)  5.040328   2.443038   2.063   0.0410 *
PC           0.140446   0.058858   2.386   0.0184 *
hsGPA       -1.802520   1.443552  -1.249   0.2140  
I(hsGPA^2)   0.337341   0.215711   1.564   0.1202  
ACT          0.004786   0.010786   0.444   0.6580  
mothcoll     0.003091   0.060110   0.051   0.9591  
fathcoll     0.062761   0.062401   1.006   0.3163  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3326 on 134 degrees of freedom
Multiple R-squared:  0.2361,	Adjusted R-squared:  0.2019 
F-statistic: 6.904 on 6 and 134 DF,  p-value: 2.088e-06


## Use the data in WAGE2.RAW for this exercise.

In [14]:
wage <- read.dta("/Users/ama/OneDrive - usp.br/S3/Natalia/Topico 5/WAGE2.DTA")

Warning message in read.dta("/Users/ama/OneDrive - usp.br/S3/Natalia/Topico 5/WAGE2.DTA"):
“cannot read factor labels from Stata 5 files”


### Holding other factors fixed, what is the approximate difference in monthly salary between blacks and nonblacks? Is this difference statistically significant?


Holding other factors fixed, black people earn about 18.8% less than nonblack people. Considering the low p-value of this coefficient, the result is very statistically significant. 

In [15]:
wage_first_lm <- lm(log(wage) ~ educ + exper + tenure + married + black + south + urban, data = wage)

summary(wage_first_lm)


Call:
lm(formula = log(wage) ~ educ + exper + tenure + married + black + 
    south + urban, data = wage)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.98069 -0.21996  0.00707  0.24288  1.22822 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept)  5.395497   0.113225  47.653  < 2e-16 ***
educ         0.065431   0.006250  10.468  < 2e-16 ***
exper        0.014043   0.003185   4.409 1.16e-05 ***
tenure       0.011747   0.002453   4.789 1.95e-06 ***
married      0.199417   0.039050   5.107 3.98e-07 ***
black       -0.188350   0.037667  -5.000 6.84e-07 ***
south       -0.090904   0.026249  -3.463 0.000558 ***
urban        0.183912   0.026958   6.822 1.62e-11 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3655 on 927 degrees of freedom
Multiple R-squared:  0.2526,	Adjusted R-squared:  0.2469 
F-statistic: 44.75 on 7 and 927 DF,  p-value: < 2.2e-16


### Add the variables `exper^2` and `tenure^2` to the equation and show that they are jointly insignificant at even the 20% level.

The F statistic for the joint significance of `exper^2` and `tenure^2` is around 1.49 with a p-value of .226. 

In [16]:
wage_squared_lm <- lm(log(wage) ~ educ + exper + I(exper^2) + tenure + I(tenure^2) + married + black + south + urban, data = wage)

summary(wage_squared_lm)


Call:
lm(formula = log(wage) ~ educ + exper + I(exper^2) + tenure + 
    I(tenure^2) + married + black + south + urban, data = wage)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.98236 -0.21972 -0.00036  0.24078  1.25127 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  5.3586756  0.1259143  42.558  < 2e-16 ***
educ         0.0642761  0.0063115  10.184  < 2e-16 ***
exper        0.0172146  0.0126138   1.365 0.172665    
I(exper^2)  -0.0001138  0.0005319  -0.214 0.830622    
tenure       0.0249291  0.0081297   3.066 0.002229 ** 
I(tenure^2) -0.0007964  0.0004710  -1.691 0.091188 .  
married      0.1985470  0.0391103   5.077 4.65e-07 ***
black       -0.1906636  0.0377011  -5.057 5.13e-07 ***
south       -0.0912153  0.0262356  -3.477 0.000531 ***
urban        0.1854241  0.0269585   6.878 1.12e-11 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3653 on 925 degrees of freedom
Multiple R-squar

In [17]:
# set the null hypothesis
null_hyp2 <- c("I(exper^2)", "I(tenure^2)")

linearHypothesis(wage_squared_lm, null_hyp2)

Res.Df RSS      Df Sum of Sq F        Pr(>F)   
1 927    123.8185 NA        NA       NA        NA
2 925    123.4210  2 0.3975639 1.489806 0.2259568

### Extend the original model to allow the return to education to depend on race and test whether the return to education does depend on race.

The coefficient of the interaction `black * educ` is -0.022, meaning that the return to education for black people is around 2.3% lower than nonblack people. However, the p-value is about 0.28, which is is not enough to reject the null hypothesis that the return to education is not dependant on race. 

In [18]:
wage_int_lm <- lm(log(wage) ~ educ + exper + I(exper^2) + tenure + I(tenure^2) + married + black + south + urban + black * educ, data = wage)

summary(wage_int_lm)


Call:
lm(formula = log(wage) ~ educ + exper + I(exper^2) + tenure + 
    I(tenure^2) + married + black + south + urban + black * educ, 
    data = wage)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.97978 -0.21976  0.00099  0.23457  1.25593 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  5.3377228  0.1273864  41.902  < 2e-16 ***
educ         0.0658997  0.0064872  10.158  < 2e-16 ***
exper        0.0173273  0.0126131   1.374 0.169849    
I(exper^2)  -0.0001283  0.0005320  -0.241 0.809527    
tenure       0.0247289  0.0081310   3.041 0.002422 ** 
I(tenure^2) -0.0007816  0.0004712  -1.659 0.097489 .  
married      0.1981442  0.0391086   5.067 4.89e-07 ***
black        0.0825126  0.2555310   0.323 0.746839    
south       -0.0898041  0.0262657  -3.419 0.000656 ***
urban        0.1853405  0.0269562   6.876 1.14e-11 ***
educ:black  -0.0218196  0.0201868  -1.081 0.280033    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


### Allow wages to differ across four groups of people: married and black, married and nonblack, single and black, and single and nonblack. What is the estimated wage differential between married blacks and married nonblacks?

The estimated wage difference between married black people and married nonblack peole is -0.179467, meaning that a married black person earns about 18% less than a married nonblack person. 

In [19]:
0.009448 - 0.188915 # coefficients from the regression below

[1] -0.179467

In [20]:
# where single and nonblack is the base group 

wage_dummies <- wage %>%
mutate(marr_nb = ifelse(married == 1 & black == 0, 1, 0),
      sin_b = ifelse(married == 0 & black == 1, 1, 0),
      marr_b = ifelse(married == 1 & black == 1, 1, 0))

In [21]:
wage_dummies_lm <- lm(log(wage) ~ educ + exper + tenure + south + urban + marr_nb + sin_b + marr_b, data = wage_dummies)

summary(wage_dummies_lm)


Call:
lm(formula = log(wage) ~ educ + exper + tenure + south + urban + 
    marr_nb + sin_b + marr_b, data = wage_dummies)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.98013 -0.21780  0.01057  0.24219  1.22889 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept)  5.403793   0.114122  47.351  < 2e-16 ***
educ         0.065475   0.006253  10.471  < 2e-16 ***
exper        0.014146   0.003191   4.433 1.04e-05 ***
tenure       0.011663   0.002458   4.745 2.41e-06 ***
south       -0.091989   0.026321  -3.495 0.000497 ***
urban        0.184350   0.026978   6.833 1.50e-11 ***
marr_nb      0.188915   0.042878   4.406 1.18e-05 ***
sin_b       -0.240820   0.096023  -2.508 0.012314 *  
marr_b       0.009448   0.056013   0.169 0.866083    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3656 on 926 degrees of freedom
Multiple R-squared:  0.2528,	Adjusted R-squared:  0.2464 
F-statistic: 39.17 on 8 and 926 D

 ## Which of the following are consequences of heteroskedasticity?
 
 (i) No. According to Wooldridge: "It is important to remember that heteroskedasticity does not cause bias or inconsistency in the OLS estimators of the $\beta_{j}$, whereas something like omitting an important variable would have this effect" (p. 263). 
 
 (ii) Yes. The usual F statistic no longer has an F distribution because  
 
 (iii) Yes. The OLS estimators are no longer BLUE.

## Consider a linear model to explain monthly beer consumption.

So $h(x) = inc^2$ where $h(x)$ is a function of the explanatory variables that determines heteroskasdisity, defined as $Var(u_{i}|x_{i}) = \sigma^2h(x_{i})$.This means we can transform the original data by dividing by $\sqrt{h_{i}}$ to get a linear regression. 

$\frac{beer}{inc} = \beta_{0} + \beta{1} + \beta_{2}(price/inc) + \beta_{3}(educ/inc) + \beta{4}(female/inc) + (u/inc)$

## True or False: WLS is preferred to OLS when an important variable has been omitted (iii) Estimate the linear probability model from the model.

False. Without further information about the relationship between the excluded variable and the other explanatory variables, it is not possible to say that WLS would have less bias than OLS. 

## Using the data in GPA3.RAW. 

### Do the variables `crsgpa`, `cumgpa`, and `tothrs` have the expected estimated effects? Which of these variables are statistically significant at the 5% level? Does it matter which standard errors are used?

This variables have the expected estimated effects. If all students tend to have higher grades in the class, then it is more likely that one student will have a higher GPA. If a student has a higher cumulative GPA prior to the semester, it is also more likely that they will have a higher term GPA, as they tend to have higher grades in general. Finally, if a student has taken more credits, and thus perhaps is more experienced, they will likely have a small but significant increase in term GPA. 

The t-statistic is the coefficient divided by its standard error. For the coefficient to be statistically significant at the 5% level, its t-statistic should be above 1.96. For `crsgpa`, the t-statistic for both usual standard error and robust standard error is statistically significant. For `cumgpa`, the t-statistic using the robust standard error is significant. However, for `tothrs`, neither standard error produces a significant t-statistic at the 5% level. 

### Why does the hypothesis $H_{0}:\beta_{crsgpa} = 1$ make sense? Test this hypothesis against the two-sided alternative at the 5% level using both standard errors. 

Without any other explanatory variables in the model, if $\beta_{crsgpa} = 1$, this means that the best predictor of a student's grade would simiply be the average GPA in the student's other classes. However, when we add the other explanatory variables, $\beta_{crsgpa}$ may not equal 1 because other variables may explain the student's term GPA. 

In both cases, we cannot reject $H_{0}:\beta_{crsgpa} = 1$ at the 5% significance level. 

In [22]:
# t statistic using the usual standard error
(0.9 - 1)/0.175

[1] -0.5714286

In [23]:
# t statistic using the robust standard error
(0.9 - 1)/0.166

[1] -0.6024096

### Is there an in-season effect on term GPA (considering both standard errors)? Does the significance level at which the null can be rejected depend on the standard error used?

Against a two-sided alternative, only the t score for using the robust standard error is just barely significant at -1.96 at the 5% level. The t score with the usual standard error does not reach the 10% level of significance at 1.65, and so is not reasonably significant. 

In [24]:
# t score with usual standard error for season
-0.157/0.098

[1] -1.602041

In [25]:
# t score with robust standard error for seasons
-0.157/0.08

[1] -1.9625

## Use the data set GPA1.RAW for this exercise.

In [26]:
gpa1_new_lm <- lm(colGPA ~ hsGPA + ACT + skipped + PC, data = gpa1)

summary(gpa1_new_lm)


Call:
lm(formula = colGPA ~ hsGPA + ACT + skipped + PC, data = gpa1)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.84006 -0.20392 -0.03352  0.25346  0.74558 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  1.35651    0.32750   4.142 6.01e-05 ***
hsGPA        0.41295    0.09243   4.468 1.65e-05 ***
ACT          0.01334    0.01044   1.278  0.20353    
skipped     -0.07103    0.02625  -2.706  0.00768 ** 
PC           0.12444    0.05731   2.171  0.03165 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3251 on 136 degrees of freedom
Multiple R-squared:  0.2593,	Adjusted R-squared:  0.2375 
F-statistic:  11.9 on 4 and 136 DF,  p-value: 2.553e-08


### Compute the special case of the White test for heteroskedasticity.

The f statistic of the White test is 3.58 with a p-value of 0.031, which means we can reject the null hypothesis and conclude that there is evidence of heteroskedasticity with confidence at the 5% level. 

In [40]:
u2 <- gpa1_new_lm$residuals^2

y <- fitted(gpa1_new_lm)

white <- lm(u2 ~ y + I(y^2))

summary(white)$fstatistic

value      numdf      dendf 
  3.581494   2.000000 138.000000

In [37]:
Ru2 <- summary(white)$r.squared

LM <- nrow(gpa1)*Ru2

p.value <- 1-pchisq(LM, 2)

p.value # 

[1] 0.03084489

### Verify that the fitted values from part (ii) are all strictly positive. Then, obtain the weighted least squares estimates. Compare the weighted least squares estimates for the effect of skipping lectures and the effect of PC ownership with the corresponding OLS estimates. What about their statistical significance?

The coefficients for skipping lectures and PC ownership are quite similar the OLS and WLS models, though both variables are slightly more statistically significant in the WLS model. 

In [45]:
sum(y < 0) # assert that all the fitted values from part (ii) are strictly positive

[1] 0

In [48]:
weights <- 1/y

gpa1_weight_lm <- lm(colGPA ~ hsGPA + ACT + skipped + PC, data = gpa1, weights = weights)

summary(gpa1_weight_lm)


Call:
lm(formula = colGPA ~ hsGPA + ACT + skipped + PC, data = gpa1, 
    weights = weights)

Weighted Residuals:
    Min      1Q  Median      3Q     Max 
-0.4834 -0.1159 -0.0210  0.1405  0.4326 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  1.37836    0.32243   4.275 3.57e-05 ***
hsGPA        0.40801    0.09096   4.486 1.53e-05 ***
ACT          0.01318    0.01032   1.278  0.20353    
skipped     -0.07197    0.02557  -2.814  0.00562 ** 
PC           0.12415    0.05712   2.173  0.03148 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.1847 on 136 degrees of freedom
Multiple R-squared:  0.262,	Adjusted R-squared:  0.2403 
F-statistic: 12.07 on 4 and 136 DF,  p-value: 1.998e-08
